Include subclasses of standard property classes as property decorators

mharding-hpe opened the pull request:
<!--
Thank you for submitting a PR to astroid!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [X] Write a good description on what the PR does.
- [X] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [X] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :sparkles: New feature |

## Description

This is gets `astroid` to automatically recognize subclasses of standard Python `property` classes (like `functools.cached_property`) as `property` decorators. This change corrects the problem I reported in [issue 10377](https://github.com/pylint-dev/pylint/issues/10377).

I also wasn't sure if I should describe this as a bug fix or a new feature. :) 

This PR also corrects an inaccuracy regarding when `enum.property` was added to Python.

<!-- If this PR fixes an issue, use the following to automatically close when we merge: -->

Refs #2306
Closes [#10377](https://github.com/pylint-dev/pylint/issues/10377)

DanielNoord reviewed (commented):
Do you mind adding a test for this? There should be other tests for `property` that we can adapt :)

mharding-hpe commented:
Actually, I have hit a problem with this PR as written. It works if my class inherits from `cached_property`, but not if it inherits from `cached_property[T]`. Will need to dig further. I'll move this PR into draft state in the meantime.

Pierre-Sassoulas reviewed (commented):
Looks great so far, thank you for digging into this !

mharding-hpe commented:
I have a possible fix to the subscripting issue. I have started working on the tests.

mharding-hpe commented:
Ok, I've updated my original fix so that it handles subclasses of `functools.cached_property`, whether or not it is subscripted. I added to the existing property tests in `test_nodes.py` to cover these cases, as well as a couple of others that were not currently being covered by that test.

mharding-hpe commented:
I ran the full test suite without my PR, and got:
```
1653 passed, 50 skipped, 15 xfailed, 1 xpassed
```

And with my PR (including the 1 additional test (`test_is_property_py312`) it created):
```
1654 passed, 50 skipped, 15 xfailed, 1 xpassed
```

mharding-hpe commented:
I see the tests don't cover a few of the non-property cases in the code. I think I know which ones I can add to address this. Hold, please.

mharding-hpe commented:
In working on filling out the code coverage, I ran into a question about how we want `pylint` to behave in a particular case.

Some context -- from looking at the `_is_property` function in `bases.py`, it is clear that `pylint` is willing to decide that a decorator is a property decorated just based on its name. For example, if the decorator is named `@cached_property` or `@reify`, then it will be assumed to be a property, even in cases where this is clearly not the case (like in the existing test, where an empty function is named `cached_property` and used successfully as a property decorator).

I am assuming that this is the desired philosophy into which my PR should fit. But I realized that my PR is making pylint pickier in some sense. Not in a regression sense -- my changes don't break any current behavior. But for example, consider the following code:

```python
import functools

class cached_property:
    pass

class user_property(cached_property):
    pass

class user_property_from_functools(functools.cached_property):
    pass

class A:
    @cached_property
    def cached_property(self): return 42
    @user_property
    def user_property(self): return 42
    @user_property_from_functools
    def user_property_from_functools(self): return 42
```

Without my PR in place, then only `A.cached_property` is considered by `pylint` to be a property.
With my PR in place, `A.cached_property` and `A.user_property_from_functools` are considered by `pylint` to be properties, but `A.user_property` is not. In other words, when I am validating the parent classes, I am not being as generous as `pylint` normally has been. This new behavior feels inconsistent, since I can use an empty class named `cached_property` and have it be considered a property, but if I subclass it, then it no longer works.

If we want to retain this idea that `pylint` will use the name as an assurance that it is a property, then I can easily change my PR so that in the above example, all of those will work. BUT, there is another case. Consider:

```python
cached_property = { int: float }
class user_property(cached_property[int]):
    pass
```

Do we want `user_property` here to also be interpreted as a property if used as a decorator? In this case, `cached_property` is not even referring to a class, but instead to a dict. I do see that `pylint` will not interpret THIS `cached_property` as a property decorator by itself, so maybe it's okay to also forbid the subclassing to be considered as such.

I'm going to modify my PR to allow all 3 cases of my first example, but I will retain the fact that it balks at that last example. Let me know if you disagree with any of that, or if I'm not being clear about it.

mharding-hpe commented:
Ok, I've added tests to cover all of the situations I described in my earlier comments. I also added tests for other things that had been missing, like for the `enum.property` class.

mharding-hpe commented:
I see that the 3.10 tests are failing with the new `enum.property` tests I added.
When looking into this failure, I noticed that `enum.property` was actually not added until Python 3.11.
I had been using checks for Python 3.10 because that it what was present in the existing `astroid` code in `bases.py`.
I think that needs to be changed to reflect when it was actually added. I can cover that with this same PR.

mharding-hpe commented:
[This PR](https://github.com/pylint-dev/astroid/pull/1028) seems to be the one which claimed `enum.properties` was added in 3.10.
[The Python docs](https://docs.python.org/3/library/enum.html#enum.property) say 3.11, which matches what I see when I try to import it under 3.10.

mharding-hpe commented:
I've added a commit to change the tests and `bases.py` to reflect that `enum.property` was added in Python 3.11, not 3.10. I'll also add the original PR author as a reviewer in case there is something going on here that is not apparent to me.

nelfin commented:
LGTM. For context, 3.10 wasn't released in Jul of 2021, that change was for
the alpha release. enum.property was removed in
https://github.com/python/cpython/pull/27010 commit b58dc58

On Thu., 15 May 2025, 8:40 am Mitch Harding, ***@***.***>
wrote:

> *mharding-hpe* left a comment (pylint-dev/astroid#2735)
> <https://github.com/pylint-dev/astroid/pull/2735#issuecomment-2881501884>
>
> I've added a commit to change the tests and bases.py to reflect that
> enum.property was added in Python 3.11, not 3.10. I'll also add the
> original PR author as a reviewer in case there is something going on here
> that is not apparent to me.
>
> Oh, I guess I don't have authority to add reviewers. Well, @nelfin
> <https://github.com/nelfin> if you happen to see this, please give your
> input.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pylint-dev/astroid/pull/2735#issuecomment-2881501884>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABSH3KPGEDQX7VMZPMAW6L26OS2XAVCNFSM6AAAAAB46YWS32VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQOBRGUYDCOBYGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mharding-hpe commented:
Aha -- that makes sense. I should have paid closer attention to the date of the original commit. Thank you for the insight!

> LGTM. For context, 3.10 wasn't released in Jul of 2021, that change was for
> the alpha release. enum.property was removed in
> https://github.com/python/cpython/pull/27010 commit b58dc58

mharding-hpe commented:
Working on this PR gave me the thought that there may be other cases that are currently overlooked in the tests (like validating the other standard property classes). @DanielNoord @Pierre-Sassoulas are you open to further PRs that are purely test improvements/additions, if I see anything worth doing?

Pierre-Sassoulas commented:
We're open to any PR making the code base better :) your work and in depth digging into issues is appreciated

DanielNoord reviewed (approved):
This looks awesome!

@Pierre-Sassoulas Do you want to review as well?

Pierre-Sassoulas commented on astroid/bases.py:
```diff
@@ -38,8 +38,9 @@
-PROPERTIES = {"builtins.property", "abc.abstractproperty"}
-if PY310_PLUS:
+PROPERTIES = {"builtins.property", "abc.abstractproperty", "functools.cached_property"}
+# enum.property was added in Python 3.11
+if PY311_PLUS:
     PROPERTIES.add("enum.property")
```
Should this be based on pylint's ``py-version`` instead ? Not sure if we have the ability to propagate this option to astroid yet. If not I don't know what exactly the design should be. (this is more a question for @DanielNoord)

DanielNoord replied:
We don't really have an option to...
Since we haven't had anybody complain about this for 2-3 years after making the mistake I wouldn't bother with adding complicated support for it.

Pierre-Sassoulas commented on tests/test_nodes.py:
```diff
@@ -924,67 +924,274 @@ def test(self):
+        ):
+            self._is_property(ast, prop)
+
+    @pytest.mark.skipif(not PY311_PLUS, reason="Uses enum.property")
```
```suggestion
    @pytest.mark.skipif(not PY311_PLUS, reason="Uses enum.property introduced in 3.11")
```

Pierre-Sassoulas commented on tests/test_nodes.py:
```diff
@@ -924,67 +924,274 @@ def test(self):
+        ):
+            self._is_property(ast, prop)
+
+    @pytest.mark.skipif(not PY311_PLUS, reason="Uses enum.property")
```
```suggestion
    @pytest.mark.skipif(not PY311_PLUS, reason="Uses enum.property introduced in 3.11")
```

Pierre-Sassoulas reviewed (commented):
This is great !
